### DISCUSSION TITLE
Running on serverless Docker (Azure Container Apps, Google Cloud Run)

### DISCUSSION THREAD

**Discussion by Zizico2** [tokio-rs/tokio#6063] [Q&A] [answered]
I realize that if I want to maximize performance I should just benchmark my different options and choose the best one for my use case, but as a rule of thumb: should I use the multi-threaded runtime when running on these sorts of server-less platforms? Where each container is getting allocated less than 1 cpu? I doubt the increased cold start time would be worth it (?).

**Comment by Noah-Kennedy** [comment, marked as answer]
Great question!

There isn't a hard and fast rule here. It's going to depend on what you're doing, what your traffic looks like, and what the environment you are running on is like.

I'd say that, in general, multi-thread is probably going to be optimal, but that you should definitely experiment on production traffic with different settings to see what gives you better results. In particular, you're going to want to tune to see how many threads is ideal if you are using multi-thread.

I will caution strongly against the use of synthetic benchmarks here. They are almost never applicable when trying to tune performance like this.

**Reply by Zizico2** [reply]
In my case, for example, I'm only allocating 0.25 of a CPU core to each ACA. Won't Tokio only ever spin up one worker thread anyway? Why not use single threaded? This could, obviously, differ if I were to allocate more CPU, but for this use case I don't think I understand why I would use the multi-threaded runtime.

**Reply by Darksonn** [reply]
If Tokio only spins up one worker thread, then a current thread runtime is probably a better fit.

**Reply by Noah-Kennedy** [reply]
Is this scaled or will it just be a single container?

**Reply by Zizico2** [reply]
It will be scaled. IDK how yet. Probably min replica of 1, scaling on CPU usage (on 70% or 80%). With some arbitrary max replicas, just to keep potential unexpected, random, usage spikes in check without massively increasing costs.

**Reply by Noah-Kennedy** [reply]
How much traffic are you expecting here? Generally I'd expect a few allocated cores per container with multi-thread to be the sweet spot.

**Reply by Zizico2** [reply]
no traffic at times, and 200 to 1000 rq/s at other times. It won't ever really get past that. ACA has a maximum of 4 CPUs per replica. AWS Fargate has a maximum of 16 CPUs now if I'm not mistaken. Google Cloud Run has a maximum of 1 CPU. 

each request is just making a request to another service and awaiting. 

Also I'm deploying a few other services that are gonna handle like 10 rq/s max.

So I guess for the first one I should allocate more cores and use multi threaded rt. Except if I end up deploying on Cloud Run (?)

**Reply by Noah-Kennedy** [reply]
I'd run some experiments here with different platforms and core counts (with 1 core using current_thread) in an A/B manner to look at what gets you the best tail latency at the lowest cost.

Don't run these experiments at separate times or dates, as traffic noise will drown out your signal. I'd recommend instead sharding your connections equally between experiment groups using something like round-robin DNS, and looking at which groups get the best tail latency (P95 at peak traffic or another critical point is often useful here) for the lowest price.

I'd recommend putting your data into grafana or similar graphs so that you can look at how things look at various points in time.

I've run a lot of experiments such as this in the past, so feel free to ask more questions.

**Reply by Zizico2** [reply]
Thanks! You've been incredibly helpful! I think I have a good plan for now. I'll bump this gain in case anything comes up!

**Reply by Noah-Kennedy** [reply]
No problem! Good luck with the project!
